### PR TITLE
Notes actions

### PR DESCRIPTION
**Pull request by AnnetaDe**

### CODE REVIEW

**Comment by liudmylailchenko** [comment]
Please update Readme.md file and add all new commands related to notes.

**Review by liudmylailchenko — APPROVED** [review]

File: src/actions.py [hunk 1]
```diff
@@ -217,3 +218,161 @@ def show_email(args, book):
...
+
+def add_note(args: list[str], book: "AddressBook", notes_book: "NoteBook") -> str:
+    """Adds a note to a contact in the address book.
+
+    param: args: List with 2 values: name and note.
+    param: book: AddressBook object to modify.
+    param: notes_book: NoteBook object to modify.
+    return: str: Result message.
+    """
+    if len(args) != 2:
```
Comment (vp-mx) [inline src/actions.py]: we do not need this as we have check in https://github.com/vp-mx/pocketpal/blob/main/src/pocketpal.py#L41
let;s remove it in separate PR

File: src/actions.py [hunk 2]
```diff
@@ -217,3 +218,161 @@ def show_email(args, book):
...
+
+    note_title, tag = args
+    if note_in_notebook := notes_book.find(note_title):
+        notes_book.add_tag(note_title, tag)
+        return f"Tag added to note -{note_in_notebook.title}."
+    return "Note not found."
+
+
+def remove_note_tag(args: list[str], notes_book: "NoteBook") -> str:
+    """Removes a tag from a note in the notes dictionary.
```
Comment (vp-mx) [inline src/actions.py:300]: After the first line of documentation should follow empty line 
https://peps.python.org/pep-0008/#documentation-strings

File: src/notes.py [hunk 3]
```diff
@@ -8,49 +8,120 @@
...
-    def __init__(self, title: str, body: str, tags: Optional[List[str]] = None) -> None:
+    def __init__(
+        self, title: str, body: str, tags: Optional[List[str]] = None, contacts: Optional[List[str]] = None
+    ) -> None:
         """Initialize the note.
-
         :param title: The title of the note.
         :param body: The body of the note.
         :param tags: A list of tags for the note.
+        :param contacts: A list of contacts if the note is attached to a contact.
```
Comment (vp-mx) [inline src/notes.py]: A set of contacts if the note is attached to a contact.

File: src/notes.py [hunk 4]
```diff
@@ -8,49 +8,120 @@
...
         """
 
         self.title = title
         self.body = body
-        self.creation_date = datetime.now()
-        self._tags = tags if tags else []
+        self.creation_date = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
```
Comment (vp-mx) [inline src/notes.py]: it better to do in representation method as you will have more flexibility to operate  with this as object. e.g. you need to get notes created after date. in this case neet to convert back to date time object, compare and convert back to string

File: src/notes.py [hunk 5]
```diff
@@ -8,49 +8,120 @@
...
+        self.tags = tags if tags else []
+        self.contacts = contacts if contacts else set()
 
     def edit(self, new_body: str) -> None:
-        """Edit the note by changing the body."""
+        """Edit the note by adding something to the body."""
+        self.body = self.body + " " + new_body
```
Comment (vp-mx) [inline src/notes.py:31]: logic implemented method looks like append rather than change

Comment (vp-mx) [reply src/notes.py:31]: shuold be smt like find text and if text in note old text = new text

File: src/notes.py [hunk 6]
```diff
@@ -8,49 +8,120 @@
...
+        """Edit the note by replacing the body."""
+        self.body = new_body
 
     def attach_to_contact(self, contact_name: str) -> None:
         """Attach the note to a contact."""
+        self.contacts.add(contact_name)
 
     def add_tag(self, tag: str) -> None:
         """Add a tag to the note."""
+        if len(tag) < 0 and len(tag) > 20:
```
Comment (vp-mx) [inline src/notes.py]: if 0 > len(tag) > 20:

File: src/notes.py [hunk 7]
```diff
@@ -8,49 +8,120 @@
...
+            raise ValueError("Tag must be between 1 and 20 characters")
+        self.tags.append(tag)
 
     def remove_tag(self, tag: str) -> None:
         """Remove a tag from the note."""
+        if tag not in self.tags:
+            raise ValueError(f"Tag {tag} not found")
```
Comment (vp-mx) [inline src/notes.py]: let's rise Helper error

File: src/notes.py [hunk 8]
```diff
@@ -8,49 +8,120 @@
...
+        self.tags.remove(tag)
+
+    def __repr__(self):
+        return f"{self.title}, {self.body}, {self.tags}, {self.contacts}"
```
Comment (vp-mx) [inline src/notes.py]: return f"{self.title}, {self.body}, {self.tags}, {self.contacts if self.contacts else ""}"

File: src/notes.py [hunk 9]
```diff
@@ -8,49 +8,120 @@
...
     def delete(self, title: str) -> None:
         """Delete a note from the notebook by title."""
+        if not title in self.data:
+            raise KeyError(f"Note with title {title} not found")
+        del self.data[title]
 
     def edit(self, title: str, new_body: str) -> None:
         """Edit the body of an existing note."""
+        if not title in self.data:
+            return KeyError(f"Note with title {title} not found")
```
Comment (vp-mx) [inline src/notes.py]: need to raise error instead of return

Comment (vp-mx) [reply src/notes.py]: HelperError

File: src/notes.py [hunk 10]
```diff
@@ -8,49 +8,120 @@
...
         """Remove a tag from a note."""
+        self.data[title].remove_tag(tag)
+
+    def show_all(self) -> List[Note]:
+        """Show all notes."""
+        return list(self.data.values())
+
+    def find(self, title: str) -> Optional[Note]:
+        """Find a note in the notebook."""
+        return self.data.get(title)
```
Comment (vp-mx) [inline src/notes.py]: return self.data.get(title) or "Note with title {title} not existed"

**Comment by vp-mx** [comment]
Cool,  🤗 a lot of new methods added. Notes module has a lot of functionality so logically it would be better to put note related functions into new module 👜

**Review by AnnetaDe — COMMENTED** [review]
almost all but readme, its in process

**Review by vp-mx — APPROVED** [review]

File: src/commands.py [hunk 11]
```diff
@@ -204,13 +216,109 @@ class Commands(Enum):
...
+        args_len=1,
+        input_help="delete-note [note]",
+        source=Source.NOTES,
+    )
+    REPLACE_NOTE = Command(
+        cli_name="replace-note",
+        description="Replaces a note.",
+        run=replace_note,
+        args_len=2,
+        input_help="replace-note [note_title] [new_body]",
```
Comment (vp-mx) [inline src/commands.py:248]: Due to rich specific syntax need to replase args from [] to <>
image

Comment (AnnetaDe) [reply src/commands.py:248]: Ок 1 min

**Pull request merged**